src/realm/sync/subscriptions.cpp

```diff
@@ -798,6 +798,47 @@ std::vector<SubscriptionSet> SubscriptionStore::get_pending_subscriptions() cons
     return subscriptions_to_recover;
 }
 
+void SubscriptionStore::notify_all_state_change_notifications(Status error)
```

Comment:
If you wanted to dedup just a little bit of code you could make this take a `StatusWith<SubscriptionSet::State>` and then do `req.promise.set_from()` on line 814. Then the part of terminate() that clears out the list of promises could just call `nofity_all_state_change_notifications(SubscriptionSet::State::Superseded)` and it would all just work.

Reply:
I thought about that and had that code initially implemented (the StatusWith is a great recommendation rather than using two optionals...), but I wasn't 100% sure if it would be appropriate to update the min_outstanding_version before the waiting on/acquiring the pending notifications lock in the call to notify. I'll look further in the code to see if it matters.

Reply:
Ah, i missed the m_min_outstanding_version bit. Yeah, then this is fine. I guess you could have a protected

`void SubscriptionStore::do_notify_all_state_change_notifications(std::unique_lock<std::mutex> lk, StatusWith<SubscriptionSet::State>)` and do a little de-duping, but this is also super nitty and probably doesn't make a big difference either way.